log/loki.go

```diff
@@ -0,0 +1,483 @@
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
+	if len(args) == 1 {
+		return h, nil
+	}
+
+	for _, arg := range args[1:] {
+		paramParts := strings.SplitN(arg, "=", 2)
+
+		if len(paramParts) != 2 {
+			return nil, fmt.Errorf("loki arguments should be in the form `address,key1=value1,key2=value2`, got %s", arg)
+		}
+
+		key, value := paramParts[0], paramParts[1]
+
+		var err error
+		switch key {
+		case "pushPeriod":
+			h.pushPeriod, err = time.ParseDuration(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki pushPeriod %w", err)
+			}
+		case "profile":
+			h.profile = true
+		case "limit":
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+			if !(h.limit > 0) {
+				return nil, fmt.Errorf("loki limit needs to be a positive number, is %d", h.limit)
+			}
+		case "msgMaxSize":
+			h.msgMaxSize, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki msgMaxSize as a number %w", err)
+			}
+			if !(h.msgMaxSize > 0) {
+				return nil, fmt.Errorf("loki msgMaxSize needs to be a positive number, is %d", h.msgMaxSize)
+			}
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.client = &http.Client{Timeout: h.pushPeriod}
+
+	go h.loop()
+	return h, nil
+}
+
+func getLevels(level string) ([]logrus.Level, error) {
+	lvl, err := logrus.ParseLevel(level)
+	if err != nil {
+		return nil, fmt.Errorf("unknown log level %s", level) // specifically use a custom error
+	}
+	index := sort.Search(len(logrus.AllLevels), func(i int) bool {
+		return logrus.AllLevels[i] > lvl
+	})
+	return logrus.AllLevels[:index], nil
+}
+
+// fill one of two equally sized slices with entries and then push it while filling the other one
+// TODO benchmark this
+//nolint:funlen
+func (h *lokiHook) loop() {
+	var (
+		msgs       = make([]tmpMsg, h.limit)
+		msgsToPush = make([]tmpMsg, h.limit)
+		dropped    int
+		count      int
+		ticker     = time.NewTicker(h.pushPeriod)
+		pushCh     = make(chan chan int64)
+	)
+
+	defer ticker.Stop()
+	defer close(pushCh)
+
+	go func() {
+		oldLogs := make([]tmpMsg, 0, h.limit*2)
+		for ch := range pushCh {
+			msgsToPush, msgs = msgs, msgsToPush
+			oldCount, oldDropped := count, dropped
+			count, dropped = 0, 0
+			cutOff := <-ch
+			close(ch) // signal that more buffering can continue
```

Comment:
I didn't mention it before, but I'm loving this double buffering approach, nice work! :+1: